man/powermand.1
\."#################################################################
\."$Id$
\."by Andrew C. Uselton <uselton2@llnl.gov> 
\."#################################################################
\."  Copyright (C) 2001-2002 The Regents of the University of California.
\."  Produced at Lawrence Livermore National Laboratory (cf, DISCLAIMER).
\."  Written by Andrew Uselton (uselton2@llnl.gov>
\."  UCRL-CODE-2002-008.
\."  
\."  This file is part of PowerMan, a remote power management program.
\."  For details, see <http://www.llnl.gov/linux/powerman/>.
\."  
\."  PowerMan is free software; you can redistribute it and/or modify it under
\."  the terms of the GNU General Public License as published by the Free
\."  Software Foundation; either version 2 of the License, or (at your option)
\."  any later version.
\."  
\."  PowerMan is distributed in the hope that it will be useful, but WITHOUT 
\."  ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or 
\."  FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License 
\."  for more details.
\."  
\."  You should have received a copy of the GNU General Public License along
\."  with PowerMan; if not, write to the Free Software Foundation, Inc.,
\."  59 Temple Place, Suite 330, Boston, MA  02111-1307  USA.
\."#################################################################
.\"
.TH PowerMan 1 "Release 1.0" "LLNL" "PowerMan"
.SH NAME
powermand \- power control and monitoring daemon
.SH SYNOPSIS
.B powermand
.I "[-options]"
.LP
.SH DESCRIPTION
.B powermand
provides power management in a data center or compute cluster environment.
.SH OPTIONS
.TP
.I "-c, --conf filename"
Override the default location of the powerman configuration file
.I /etc/powerman/powerman.conf.
.TP
.I "-f, --foreground"
Do not daemonize, and send debugging/error messages to stderr instead of syslog.
.TP
.I "-d, --debug mask"
Set mask for debugging output.
.TP
.I "-h, --help"
Provide a synopsis of the command options.
.B PowerMan
and exit.
.SH "FILES"
/usr/sbin/powermand
.br
/usr/bin/powerman
.br
/etc/powerman/powerman.conf
.SH "ORIGIN"
Developed by Andrew  Uselton <useton2@llnl.gov> on LLNL's Linux 
clusters.  This software is open source and distributed under
the terms of the Gnu GPL.  
.SH "SEE ALSO"
powerman(1) powerman.conf(5)


NEWS
=========================================================================
Release Notes for PowerMan version 2.4                        26 Nov 2008
=========================================================================

* Initial powerman client API.

* Run the powerman daemon (and all coprocesses) as 'daemon' by default
  instead of root.  To override, set USER=root or other user in
  /etc/sysconfig/powerman.

* New man pages for httppower, plmpower.

=========================================================================
Release Notes for PowerMan version 2.3                        12 Nov 2008
=========================================================================

* Make init script work on Solaris.

* IPv6 support.

* Include example powerman.conf file (/etc/powerman/powerman.conf.example)

* Added support for Cyclades PM20, PM42.

=========================================================================
Release Notes for PowerMan version 2.2                        27 Aug 2008
=========================================================================

* Send a SIGINT to coprocesses rather than just closing file
  descriptors during powermand shutdown.  This is required to 
  terminate an ssh client running in coprocess mode.

* Make it possible to only define the _ranged version of scripts.  Scripts 
  are selected using the following precedence: _all, _ranged, singlet.

* Drop "soft power status" support.  This allowed iceboxes to detect 
  whether a node was powered up or not, independently of plug state.

* [ipmipower] Drop _all and singlet version of scripts. 

* [ilom] Revised Sun Integrated Lights Out Management support 
  to work over ssh and serial.  Dropped shared console/serial 
  "ilom-inline" support.

* [icebox3] Now supports both v3 and v4 Ice Box.  Users of icebox4
  should change their device types to icebox3 in powerman.conf.

* [plmpower] Added support for controlling Insteon/X10 devices via 
  SmartLabs PLM 2412S.

* [hp3488] Added support for modular GPIB-based HP 3488A switch/control 
  unit via gpib-utils project.

* [ics8064] Added support for VXI-11-based ICS 8064 16-port relay unit
  via gpib-utils project.

* [powerman] Improved efficiency of powerman-controlling-powerman 
  configurations, when one powerman controls a subset of the plugs of 
  another powerman through the use of _ranged scripts.

* Enhanced integrated test suite.
  Note: tests all pass on AIX now.

=========================================================================
Release Notes for PowerMan version 2.1                        18 Jun 2008
=========================================================================

* Client overhaul.

* [apcpdu3] Fix intermittent query failures.

* Fixed off by one bug in server that allowed some delays in dev 
  scripts to take longer than programmed.

* Use raw mode in pipe (|&) connections to avoid local tty echo.

* Run server in /var/run/powerman rather than /etc/powerman.
  Create this directory if it doesn't exist.

* Enhanced integrated test suite.
  Note: all tests do not currently pass on AIX.

=========================================================================
Release Notes for PowerMan version 2.0                        01 Jun 2008
=========================================================================

* Autoconf/automake integration.

* Integrated test suite.

* Portability to Solaris, AIX, and OS/X.

* Support for selecting power control targets using genders (-g option).

* Cleanup and refactoring.

=========================================================================
Release Notes for PowerMan version 1.0.32                     24 Jan 2008
=========================================================================

* Support for new apc firmware (apcpdu3.dev) [Trent D'Hooge]

* Added httppower utility for interfacing to power controllers that are
  exclusively web-based.

* Support for Digital Loggers, Inc. power controllers.

* Support for APPRO power controller.

=========================================================================
Release Notes for PowerMan version 1.0.31                     20 Oct 2007
=========================================================================

(Includes changes in 1.0.29 and 1.0.30)

* Support SUN ILOM inline with serial console, i.e.
  device "ilom25" "ilom-inline" "/usr/bin/conman -Q -j ilc25 |&"

* Handle config files with embedded carriage returns [Trent D'Hooge
  and Todd Heer].

* Ipmipower device script tuning [Al Chu].

* Minor build/packaging tweaks for building under mock, etc.

=========================================================================
Release Notes for PowerMan version 1.0.28                     07 May 2007
=========================================================================

* Increase "cycle" delays from 2s to 4s on all devices that implement
  cycle as an explicit on,delay,off.

=========================================================================
Release Notes for PowerMan version 1.0.27                     24 Apr 2007
=========================================================================

* Add heartbeat stonith module.

* Fix bug in baytech-rpc28-nc device support that affected plugs > 10.

=========================================================================
Release Notes for PowerMan version 1.0.26                     21 Dec 2006
=========================================================================

* Support 8-port APC AP7900 and likely AP7901, AP7920, and AP7921.
  [Martin Petersen]

* Work around a bug that causes baytech rpc3's to resturn some plugs
  as status "unknown" on login timeout/reconnect.

* Add ranged power control support for faster power control w/ ipmipower.

=========================================================================
Release Notes for PowerMan version 1.0.25                     16 Aug 2006
=========================================================================


powermand/powermand.c
/*****************************************************************************
 *  Copyright (C) 2001 The Regents of the University of California.
 *  Produced at Lawrence Livermore National Laboratory (cf, DISCLAIMER).
 *  Written by Andrew Uselton <uselton2@llnl.gov>
 *  UCRL-CODE-2002-008.
 *
 *  This file is part of PowerMan, a remote power management program.
 *  For details, see http://code.google.com/p/powerman/
 *
 *  PowerMan is free software; you can redistribute it and/or modify it under
 *  the terms of the GNU General Public License as published by the Free
 *  Software Foundation; either version 2 of the License, or (at your option)
 *  any later version.
 *
 *  PowerMan is distributed in the hope that it will be useful, but WITHOUT
 *  ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 *  FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
 *  for more details.
 *
 *  You should have received a copy of the GNU General Public License along
 *  with PowerMan; if not, write to the Free Software Foundation, Inc.,
 *  59 Temple Place, Suite 330, Boston, MA  02111-1307  USA.
\*****************************************************************************/

#if HAVE_CONFIG_H
#include "config.h"
#endif

#include <errno.h>
#include <sys/time.h>
#include <time.h>
#if HAVE_GETOPT_H
#include <getopt.h>
#endif
#include <unistd.h>
#include <signal.h>
#include <string.h>
#include <stdlib.h>
#include <stdio.h>
#include <assert.h>
#include <limits.h>
#include <stdarg.h>

#include "xtypes.h"
#include "list.h"
#include "hostlist.h"
#include "parse_util.h"
#include "xmalloc.h"
#include "xpoll.h"
#include "xsignal.h"
#include "pluglist.h"
#include "device.h"
#include "daemon.h"
#include "client.h"
#include "error.h"
#include "debug.h"
#include "hprintf.h"
#include "powerman.h"

/* prototypes */
static void _usage(char *prog);
static void _version(void);
static void _noop_handler(int signum);
static void _exit_handler(int signum);
static void _select_loop(void);

#define OPTIONS "c:fhd:VsY1"
#if HAVE_GETOPT_LONG
#define GETOPT(ac,av,opt,lopt) getopt_long(ac,av,opt,lopt,NULL)
static const struct option longopts[] = {
    {"conf",            required_argument,  0, 'c'},
    {"foreground",      no_argument,        0, 'f'},
    {"help",            no_argument,        0, 'h'},
    {"debug",           required_argument,  0, 'd'},
    {"version",         no_argument,        0, 'V'},
    {"stdio",           no_argument,        0, 's'},
    {"short-circuit-delay", no_argument,    0, 'Y'},
    {"one-client",      no_argument,        0, '1'},
    {0, 0, 0, 0}
};
#else
#define GETOPT(ac,av,opt,lopt) getopt(ac,av,opt)
#endif

int main(int argc, char **argv)
{
    int c;
    char *config_filename = NULL;
    bool daemonize = TRUE;
    bool use_stdio = FALSE;
    bool short_circuit_delay = FALSE;
    bool one_client = FALSE;

    /* parse command line options */
    err_init(argv[0]);
    while ((c = GETOPT(argc, argv, OPTIONS, longopts)) != -1) {
        switch (c) {
        case 'Y': /* --short-circuit-delay */
            short_circuit_delay = TRUE;
            break;
        case 'c': /* --conf */
            if (!config_filename)
                config_filename = xstrdup(optarg);
            break;
        case 'f': /* --foreground */
            daemonize = FALSE;
            break;
        case 'd': /* --debug */
            {
                unsigned long val = strtol(optarg, NULL, 0);

                if ((val == LONG_MAX || val == LONG_MIN)
                    && errno == ERANGE)
                    err_exit(TRUE, "strtol on debug mask");
                dbg_setmask(val);
            }
            break;
        case 'V': /* --version */
            _version();
            /*NOTREACHED*/
            break;
        case 's': /* --stdio */
            use_stdio = TRUE;
            one_client = TRUE;
            break;
        case '1': /* --one-client */
            one_client = TRUE;
            break;
        case 'h': /* --help */
        default:
            _usage(argv[0]);
            /*NOTREACHED*/
            break;
        }
    }

    if (use_stdio && daemonize)
        err_exit(FALSE, "--stdio should only be used with --foreground");

    if (!config_filename)
        config_filename = hsprintf("%s/%s/%s", X_SYSCONFDIR,
                                   "powerman", "powerman.conf");

    dev_init(short_circuit_delay);
    cli_init();

    conf_init(config_filename);
    xfree(config_filename);

    xsignal(SIGHUP, _noop_handler);
    xsignal(SIGTERM, _exit_handler);
    xsignal(SIGINT, _exit_handler);
    xsignal(SIGPIPE, SIG_IGN);

    cli_start(use_stdio, one_client);

    if (daemonize) {
        char *rundir = hsprintf("%s/run/powerman", X_LOCALSTATEDIR);
        char *pidfile = hsprintf("%s/powermand.pid", rundir);
        int *fds, len;

        cli_listen_fds(&fds, &len);
        daemon_init(fds, len, rundir, pidfile, DAEMON_NAME);
        xfree(rundir);
        xfree(pidfile);
        err_notty();
        dbg_notty();
    }

    /* We now have a socket at listener fd running in listen mode */
    /* and a file descriptor for communicating with each device */
    _select_loop();
    return 0;
}

static void _usage(char *prog)
{
    printf("Usage: %s [OPTIONS]\n", prog);
    printf("  -c --conf <filename>   Specify config file path\n");
    printf("  -d --debug <mask>      Set debug mask [0]\n");
    printf("  -f --foreground        Don't daemonize\n");
    printf("  -V --version           Report powerman version\n");
    printf("  -s --stdio             Talk to client on stdin/stdout\n");
    printf("  -1 --one-client        Terminate when client disconnects\n");
    exit(0);
}

static void _version(void)
{
    printf("%s\n", VERSION);
    exit(0);
}

static void _select_loop(void)
{
    struct timeval tmout;
    xpollfd_t pfd = xpollfd_create();

    timerclear(&tmout);

    /* start non-blocking connections to all the devices - finish them inside
     * the poll loop.
     */
    dev_initial_connect();

    while (1) {
        int n;

        xpollfd_zero(pfd);

        cli_pre_poll(pfd);
        dev_pre_poll(pfd);

        n = xpoll(pfd, timerisset(&tmout) ? &tmout : NULL);
        timerclear(&tmout);

        /*
         * Process activity on client and device fd's.
         * If a device requires a timeout, for example to reconnect or
         * to process a scripted delay, tmout is updated.
         */
        cli_post_poll(pfd);
        dev_post_poll(pfd, &tmout);

        if (cli_server_done())
            break;
    }
    xpollfd_destroy(pfd);
}

static void _noop_handler(int signum)
{
    /* do nothing */
}

static void _exit_handler(int signum)
{
    cli_fini();
    dev_fini();
    conf_fini();
    err_exit(FALSE, "exiting on signal %d", signum);
}

/*
 * vi:tabstop=4 shiftwidth=4 expandtab
 */
